docs/source/examples/qlora_finetune.rst

```diff
@@ -0,0 +1,283 @@
+=============================
+
+In this tutorial, we'll learn about `QLoRA <https://arxiv.org/abs/2305.14314>`_, an enhancement on top of
+`LoRA <https://arxiv.org/abs/2106.09685>`_ that maintains frozen model parameters in 4-bit quantized precision, thereby reducing memory usage. We'll
+walk through how QLoRA can be utilized within TorchTune to finetune a Llama2-7b model in < 10 GB of memory.
+It is highly recommended to first develop an understanding of :ref:`LoRA finetuning in TorchTune<lora_finetune_label>`.
+
+
+.. grid:: 2
+
+    .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
+
+      * How QLoRA saves memory over LoRA finetuning
+      * An overview of QLoRA in TorchTune
+      * How to run a QLoRA finetune in TorchTune
+
+    .. grid-item-card:: :octicon:`list-unordered;1em;` Prerequisites
+
+      * Be familiar with :ref:`TorchTune<overview_label>`
+      * Make sure to :ref:`install TorchTune<install_label>`
+      * Make sure you have downloaded the :ref:`Llama2-7B model weights<download_llama_label>`
+      * Be familiar with :ref:`LoRA in torchtune<lora_finetune_label>`
+
+What is QLoRA?
+---------------
+
+`QLoRA <https://arxiv.org/abs/2305.14314>`_ builds on top of `LoRA <https://arxiv.org/abs/2106.09685>`_ to enable additional
+memory efficiency on top of LoRA. In LoRA, model parameters can be thought of as existing in two partitions: adapters, which are
+low-rank matrices added to different layers of a neural network, and base model parameters, which are parameters that are part of
+the original model. In vanilla LoRA style training, both these parameters are held in the same precision (typically fp32 or bf16), and
+therefore activations and intermediate gradients computed are in fp32/bf16.
+
+QLoRA further quantizes the base model parameters into a bespoke 4-bit NormalFloat (NF4) data type, resulting in 4x less parameter memory usage while
+largely retaining model accuracy. As a result, the vast majority of parameters only take up 4 bits (as opposed to 16 or 32 bits by bf16/fp32 dtypes). This
+quantization is done through the method highlighted in the original `QLoRA paper <https://arxiv.org/abs/2305.14314>`_. Adapter
+parameters are still held in the original precision, and activations, gradients, and optimizer states still exist in the higher precision to preserve
+accuracy.
+
+The `QLoRA paper <https://arxiv.org/abs/2305.14314>`_ introduces two key abstractions to decrease memory usage and avoid accuracy degradation: the bespoke 4-bit NormatFloat
+type, and a double quantization method that quantizes the quantization parameters themselves to save even more memory. TorchTune uses
+the `NF4Tensor abstraction <https://github.com/pytorch-labs/ao/blob/b9beaf351e27133d189b57d6fa725b1a7824a457/torchao/dtypes/nf4tensor.py#L153>`_ abstraction from the `TorchAO library <https://github.com/pytorch-labs/ao>`_ to build QLoRA components as specified in the paper.
+
+
+QLoRA Core Abstractions and Usage
```

Comment:
I have mixed feelings about this section, I think framing it as diving into core abstractions is a bit too technical for a tutorial and a lot of the code might be too much information. Most users just want a high level understanding of what QLoRA is and how they can use it in torchtune. I think the example code showing the cuda memory allocated differences is awesome, and maybe you can keep a demo of the linear_nf4 but in a more contrived example that shows that the quantized weight is saved for backward.

Or maybe another way to organize it: high level of what is QLoRA, how do you run it in torchtune, show the memory differences, and use the memory allocated / linear_nf4 examples to demonstrate how memory is saved. This way, you introduce the concepts from high level and easy to understand to more specific and more nitty gritty.

Reply:
So I mostly agree, though I do think tutorials, in addition to clearly showcasing how the feature can be used in TorchTune, can also add value by deep diving into how things work under the hood to help with user understanding. If users purely want to use QLoRa for a finetune, they can do so at a high level via the README. 

For this particular section, I tried to address this concern by having the first portion of this section be pretty high level - how to just pass in `quantize_base` flag to LoRALinear and get a QLoRA enabled linear. The rest of the section goes deeper into how it actually happens - maybe can add a "this part is optional" flag or something like that.

Reply:
I think the optional part makes sense here. 

The other thing I would recommend just in terms of format is just to switch it up between code blocks and text (and if possible images) a bit, especially on the denser topics. Maybe I have a short attention span, but once I read >=3 sentences of upcast/downcast/dtype talk my eyes start to glaze over, and changing the medium you're conveying information with is a good way to keep things engaging. 

Also can be worth thinking about if there are places where it's OK to be less than 100% precise for the sake of getting the essential idea across, then point to a reference elsewhere. (Admittedly I already looked through here for places to do that, and I think the subject matter just makes it a bit hard tbh.. it's often either all the gory details or none at all)

Reply:
Don't like flagging a whole section as optional, that feels odd to me. How about just changing the order of these sections and making this section a little more succinct? Because the meat is how do we run QLoRA in torchtune, and that should be higher up.

Reply:
Can we move this to a Deep Dive article?

Reply:
Ok, I moved this to the bottom right now to focus on using QLoRA in torchtune and leaving this as a deep dive if folks are interested. Can move it to a separate article in a follow up PR if that's an even better approach.